Add missing input-handling functions.

Kevinpgalligan opened the pull request:
Including: on-click-release, on-middle-click-release, on-right-click-release, on-text and on-button.

Also, add a call to on-hover for the window, not just entities.

Also, update the docs & examples.

Relevant issue: https://github.com/vydd/sketch/issues/125

vydd commented on src/controllers.lisp:
```diff
@@ -47,6 +49,10 @@
       (call-next-method))))
 
+(defmethod on-hover :around ((*sketch* sketch) x y)
+  (with-sketch (*sketch*)
```
*draw-mode* nil missing - also, shouldn't we propagate-to-entity like everywhere else?

vydd commented on src/controllers.lisp:
```diff
@@ -16,11 +16,13 @@
 (defmethod on-click (instance x y))
 (defmethod on-middle-click (instance x y))
 (defmethod on-right-click (instance x y))
+(defmethod on-click-release (instance x y))
```
on-release?

vydd commented on src/controllers.lisp:
```diff
@@ -99,5 +108,29 @@
 
-(defmethod keyboard-event :after ((instance sketch)
-                                  state timestamp repeatp keysym))
+(defconstant +scancode-prefix-length+ 9)
```
could we change this to

```lisp
(defconstant +scancode-prefix-length+ (length "scancode-"))
```

? thanks!

vydd commented on src/controllers.lisp:
```diff
@@ -99,5 +108,29 @@
+
+(defmethod on-text :around ((*sketch* sketch) text)
+  (with-sketch (*sketch*)
+    (call-next-method)))
```
draw mode missing

vydd commented on src/controllers.lisp:
```diff
@@ -99,5 +108,29 @@
+
+(defmethod on-key :around ((*sketch* sketch) key state)
+  (with-sketch (*sketch*)
+    (call-next-method)))
```
draw mode missing

vydd commented on src/controllers.lisp:
```diff
@@ -16,11 +16,13 @@
 (defmethod on-middle-click (instance x y))
 (defmethod on-right-click (instance x y))
+(defmethod on-click-release (instance x y))
+(defmethod on-middle-click-release (instance x y))
```
are the new methods not using `with-sketch` / `draw-mode` ?

probably a good candidate for a file-local macro..but we can do that later

Kevinpgalligan replied:
I think I forgot to add that.

Kevinpgalligan replied:
And yes, a macro would be nice to remove all the repetitive code.

vydd commented on src/controllers.lisp:
```diff
@@ -16,11 +16,13 @@
 (defmethod on-middle-click (instance x y))
 (defmethod on-right-click (instance x y))
+(defmethod on-click-release (instance x y))
+(defmethod on-middle-click-release (instance x y))
```
would it make sense to add `-press` methods as well?

Kevinpgalligan replied:
Hmm, I was intuitively interpreting `on-click` to mean "on press". But now that you say it, onclick events in JavaScript are triggered when the click is released. So it might make more sense that way.

Kevinpgalligan commented:
I think I've addressed all the feedback!

- Set draw-mode to nil where appropriate.
- Propagate events to entities.
- Replace repetitive method definitions with macro.
- Remove magic number.
- Change the interface so that `on-click` refers to a full click, including the release of the mouse button, as is the case e.g. in JavaScript. Now there's `on-click-press` for the initial press of the mouse button. Open to changing this if there are alternative suggestions.

vydd commented on src/controllers.lisp:
```diff
@@ -57,14 +56,17 @@
 (defmethod kit.sdl2:mousebutton-event ((instance sketch) state timestamp button x y)
   (let ((button (elt (list nil :left :middle :right) button))
-        (method (elt (list nil #'on-click #'on-middle-click #'on-right-click) button)))
+        (click-method (elt (list nil #'on-click-press #'on-middle-click-press #'on-right-click-press) button))
```
hey! thanks for making the changes! I'm confused here now --- it's saying click methods are the press ones, and release methods are .. click?

why don't we do all three methods - press, click, release? I feel like release is interesting in a drag  & drop scenario

Kevinpgalligan replied:
Click methods are release, ya. That's how it works in JavaScript (`onclick()`) and I think it matches people's intuitions. E.g. if you press down the mouse button and move it around, you don't expect the click to take effect until you release the button. Could add an explicit `on-click-release` but it would be doing the same thing as `on-click`.
